### PR TITLE
Add new stages and images to the Dockerfile

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

We are getting to the point where we want users (non-developers) to be able to look at SpECTRE and run a few things. Our current Docker image on DockerHub doesn't make this very easy for people. To alleviate this, I have edited the Dockerfile so we have different images for different uses.

The new images will be different tags of a new DockerHub repo `sxscollaboration/spectre`. The three images are:

1. `sxscollaboration/spectre:dev`:  The bare minimum needed to compile SpECTRE. Good for development purposes.
2. `sxscollaboration/spectre:ci`:  Add extra libraries on top of base which are needed to do testing on GitHub CI
3. `sxscollaboration/spectre:demo`: Adds Paraview and has pre-built executables `ExportTimeDependentCoordinates3D` and `EvolveScalarAdvectionKuzmin2D` so users don't have to worry about compiling SpECTRE in order to run a few test problems.

Currently, only the `dev` and `ci` images are on DockerHub. A future PR will edit the `PostRelease` workflow to push the updated pre-built executables to the `demo` image. Also, the `demo` image will also have the `latest` tag which is Docker's default tag for repos. So after the future PR, the default behavior for

```
docker pull sxscollaboration/spectre
```

will be to pull the `demo` image so people can have a clone of the SpECTRE repo and the pre-built executables to play around with.

Thank you to @nilsdeppe for building and pushing the `dev` and `ci` images so this PR can actually be tested on CI.

### Upgrade instructions

Once this is merged, rebase on the latest develop branch to use the new container for CI on your PR.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: containers/Dockerfile.buildenv
Comment (nilsdeppe) [inline containers/Dockerfile.buildenv]: I guess this isn't true/needed since we don't have yapf for python2 anymore?

**Review by nilsdeppe — COMMENTED** [review]
LGTM, I've pushed the latest container. Just one comment to test out to simplify maintenance if possible. Squash right away if it works, and if not don't worry about it :)

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv]: Add `(in the container)`

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv]: `:demo`

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv:212]: Why this change?

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv]: inherit

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv]: build

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv:341]: There's a way to fetch the image and then make sure that Docker uses it as cache when building. Maybe add a note here to try this?

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv]: Also build all-pybindings I think, add bin/python to PYTHONPATH, and `pip install jupyterlab` for demos/tutorials.

**Review by nilsvu — COMMENTED** [review]
Nice work, thanks for doing this 👍

File: containers/Dockerfile.buildenv
Comment (knelli2) [reply containers/Dockerfile.buildenv:212]: For some reason, it would error if I left it as `cd /work`. Didn't dig into why

File: containers/Dockerfile.buildenv
Comment (knelli2) [reply containers/Dockerfile.buildenv]: Deleted `futures`, rebuilt the container, and then built unit tests and python bindings with python2 and everything worked fine so I think I can get rid of it.

File: containers/Dockerfile.buildenv
Comment (knelli2) [reply containers/Dockerfile.buildenv:341]: Is that not what this does? Pulls the remote image and uses that as cache and builds `demo` on top of it?

If it isn't, could you refer me to what you had in mind?

**Comment by knelli2** [comment]
Pushed a fixup addressing requested changes.
- Tested out that I was able to run `jupyter lab` inside the demo image and view it in chrome
- Removed `futures` from python2 as it was only needed for py2 yapf (which I got rid of)
- Added a `.dockerignore` file in SPECTRE_ROOT that ignores all directories named `build*` just like `.gitignore`. This way you don't have to worry about your build directories being copied from the docker context. (Also had to add this to formaline)

Edit: @nilsdeppe you'll probably have to rebuild the images and restart CI after I squash

File: containers/Dockerfile.buildenv
Comment (nilsvu) [reply containers/Dockerfile.buildenv:341]: You would normally base `demo FROM dev`, not from the remote image that must be pushed before. My understanding is you based on the remote image so the post-release workflow doesn't have to rebuild `dev`. And my comment is: there's a way to pre-fetch `dev` so Docker doesn't need to rebuild it. My suggestion is to add a comment so it's clear this is something we could try, but didn't bother just yet.

Comment (knelli2) [reply containers/Dockerfile.buildenv:341]: Ah ok I understand. I added in a couple lines to the fixup explaining this.

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv:279]: You can also move the Intel SDE installation to the CI container

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv]: What is this line for? Did you mean to point to /opt/paraview?

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv]: No need to update apt

**Review by nilsvu — DISMISSED** [review]

File: containers/Dockerfile.buildenv
Comment (knelli2) [reply containers/Dockerfile.buildenv]: No, I did have `/opt/paraview` before, but it messed up cmake finding python when we built spectre, even when I gave it the correct python executable. I think I can remove this line.

Comment (nilsvu) [reply containers/Dockerfile.buildenv]: Ok installing ParaView pybindings can be tricky. I can give it a try later.

File: containers/Dockerfile.buildenv
Comment (nilsvu) [inline containers/Dockerfile.buildenv]: Everything

**Review by nilsvu — DISMISSED** [review]

File: containers/Dockerfile.buildenv
Comment (knelli2) [reply containers/Dockerfile.buildenv]: Why am I so bad at spelling 😓

Comment (nilsvu) [reply containers/Dockerfile.buildenv]: https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker 😉

Comment (knelli2) [reply containers/Dockerfile.buildenv]: What makes it even worse is that I actually have that extension enabled and still missed it. 😅  (To be fair, it got lost in all of the "incorrect" spellings of library names and bash commands)

**Comment by nilsdeppe** [comment]
I'm happy with this. @nilsvu please let us know when @knelli2 can squash and if we should wait for your explicit re-approval post-squash :)

**Comment by nilsvu** [comment]
Oh I already approved with the comment above, so feel free to squash and don't wait for me with merging this 👍

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**
